walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,66 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Base cases
+    if (request.getAttendees().isEmpty()) { return new ArrayList<>(Arrays.asList(TimeRange.WHOLE_DAY)); }
+    if (request.getDuration() >= TimeRange.WHOLE_DAY.duration()) { return new ArrayList<>(); }
+    
+    // Preprocess event input
+    List<Event> pertinentEvents = new ArrayList<>(filterEvents(new ArrayList<>(events), request.getAttendees()));
+    Collections.sort(pertinentEvents, (Event e1, Event e2) -> TimeRange.ORDER_BY_END.compare(e1.getWhen(), e2.getWhen()));
+    
+    List<TimeRange> openSlots = new ArrayList<>();
+    int previousEnd = 0;
+
+    for (Event event : pertinentEvents) {
+      TimeRange eventTimeRange = event.getWhen();
+      if (eventTimeRange.start() - previousEnd >= request.getDuration()) {
+        openSlots.add(TimeRange.fromStartEnd(previousEnd, eventTimeRange.start(), false));
+      }
+      previousEnd = eventTimeRange.end();
+    }
+    if ((previousEnd < TimeRange.END_OF_DAY) && (TimeRange.END_OF_DAY - previousEnd >= request.getDuration())) {
```

Comment:
I think the first part of this if statement is unnecessary --asssume `previousEnd is not < TimeRange.END_OF_DAY`, then `end of day - previous end` will not be greater than or equal to request duration because it'll be negative

Reply:
Ah that's true, done